From ba5c1ac5527aca70ded4fbbbc466c123d2b7e9d3 Mon Sep 17 00:00:00 2001
From: Mingcong Bai <jeffbai@aosc.io>
Date: Wed, 26 Feb 2025 11:43:02 +0800
Subject: [PATCH 24/24] AOSCOS: USB: core: only enable root_hub wakeup on
 MACH_LOONGSON64

The previous attempt to fix USB remote wake on Loongson 7A platforms broke
S3 suspend on at least the following device(s):

- Lenovo ThinkPad P14s Gen 1 (20Y1CTO1WW)

Where the device would wake up immediately after entering S3 suspend.
Since this fix was really only intended as a quirk for Loongson 7A
platforms, guard the additional calls to `usb_enable_remote_wakeup()' and
`usb_disable_remote_wakeup()' within `CONFIG_MACH_LOONGSON64',
corresponding to all 64-bit Loongson platforms (unless this also breaks
non-7A devices, in which case we will tighten the conditions further).

Fixes: "FROMLIST: USB: core: Enable root_hub's remote wakeup for wakeup sources"
Signed-off-by: Mingcong Bai <jeffbai@aosc.io>

Link: https://t.me/c/1109254909/816220
Link: https://t.me/c/1109254909/820298
Signed-off-by: Kexy Biscuit <kexybiscuit@aosc.io>
---
 drivers/usb/core/hub.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/usb/core/hub.c b/drivers/usb/core/hub.c
index 390f82b48..5bf1bd87a 100644
--- a/drivers/usb/core/hub.c
+++ b/drivers/usb/core/hub.c
@@ -3522,7 +3522,9 @@ int usb_port_suspend(struct usb_device *udev, pm_message_t msg)
 			if (PMSG_IS_AUTO(msg))
 				goto err_wakeup;
 		}
+#if defined(CONFIG_MACH_LOONGSON64)
 		usb_enable_remote_wakeup(udev->bus->root_hub);
+#endif
 	}
 
 	/* disable USB2 hardware LPM */
@@ -3588,7 +3590,9 @@ int usb_port_suspend(struct usb_device *udev, pm_message_t msg)
 
 		if (udev->do_remote_wakeup) {
 			(void) usb_disable_remote_wakeup(udev);
+#if defined(CONFIG_MACH_LOONGSON64)
 			(void) usb_disable_remote_wakeup(udev->bus->root_hub);
+#endif
 		}
  err_wakeup:
 
-- 
2.51.2

